### PR TITLE
Add fusion patterns for conformer-transducer model

### PR DESCRIPTION
**Pull request by apsonawane**
### Description
Add conformer-transducer model type to optimizer. This PR adds pattern matches for attention shown below:
Unfused attention:
ct_unfused

Fused attention:
ct_fused

### CODE REVIEW

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 1]
```diff
@@ -692,13 +696,27 @@ def create_multihead_attention_node(
...
+            and present_k in graph_output_names
+            and present_v in graph_output_names
+        ):
+            kv_cache_name_match = True
+        else:
+            kv_cache_name_match = False
+
         # Add optional inputs for MHA
-        if past_k and past_v and past_k in graph_input_names and past_v in graph_input_names:
+        if past_k and past_v and kv_cache_name_match:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_attention.py:710]: This logic change 
```
        if past_k and past_v and past_k in graph_input_names and past_v in graph_input_names:
            mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
   ```
   to 
   ```
     if past_k and past_v:
            mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
   ```
   might impact existing model. Could you only change the logic for CT model, but not impact other model?

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 2]
```diff
@@ -692,13 +696,27 @@ def create_multihead_attention_node(
...
+            mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
+        elif past_k and past_v and not kv_cache_name_match:
             mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
 
         # Add outputs for MHA
         mha_outputs = [output]
-        if present_k and present_v and present_k in graph_output_names and present_v in graph_output_names:
+        if present_k and present_v and kv_cache_name_match:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_attention.py:717]: Same as above comment.
```
        if present_k and present_v and kv_cache_name_match:
            mha_outputs.extend([present_k, present_v])
        elif present_k and present_v and not kv_cache_name_match:
            mha_outputs.extend([present_k, present_v])
```
is equal to

```
        if present_k and present_v:
            mha_outputs.extend([present_k, present_v])
```

You may use some logic like 
```
 if present_k and present_v and (present_k in graph_output_names and present_v in graph_output_names  or is_ct_model):
```

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 3]
```diff
@@ -692,13 +696,27 @@ def create_multihead_attention_node(
...
+            and present_v in graph_output_names
+        ):
+            kv_cache_name_match = True
+        else:
+            kv_cache_name_match = False
+
         # Add optional inputs for MHA
-        if past_k and past_v and past_k in graph_input_names and past_v in graph_input_names:
+        if past_k and past_v and kv_cache_name_match:
+            mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_attention.py:711]: Although not introduced it this PR, the logic is questionable. 
why appending add_qk input depends on past_k and past_v is graph inputs?

**Comment by tianleiwu** [comment]
Pease add a test case for the attention fusion. Otherwise, it is not able to prevent regression in the future.

File: onnxruntime/python/tools/transformers/fusion_attention.py (hunk 1 above)
Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_attention.py:710]: Makes sense will update with the suggested logic. One other option is to write same function for the CT model but that will have lot of duplicate code

File: onnxruntime/python/tools/transformers/fusion_attention.py (hunk 3 above)
Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_attention.py:711]: It should not depend on past_k and past_v values that should be added separately, also key_padding_mask should be separate right?

File: onnxruntime/python/tools/transformers/optimizer.py [hunk 4]
```diff
@@ -56,6 +57,7 @@
     "unet": (UnetOnnxModel, "pytorch", 1),  # UNet in Stable Diffusion
     "vae": (VaeOnnxModel, "pytorch", 1),  # UAE in Stable Diffusion
     "vit": (BertOnnxModel, "pytorch", 1),
+    "ct": (TransducerOnnxModel, "pytorch", 1),
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/optimizer.py:60]: AFAIK, current optimization is mainly the conformer part of CT model. Right?
image

If so, I suggest to rename CT as conformer. TransducerOnnxModel as ConformerOnnxModel.

Comment (apsonawane) [reply onnxruntime/python/tools/transformers/optimizer.py:60]: Yes the fusion is for conformer, will update it to ConformerOnnxModel

File: onnxruntime/python/tools/transformers/fusion_attention.py (hunk 3 above)
Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_attention.py]: I updated this but whisper tests are failing because of this. Will address this issue in another PR since it is unrelated to this one

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 5]
```diff
@@ -693,12 +692,13 @@ def create_multihead_attention_node(
             mha_inputs.append("")
 
         # Add optional inputs for MHA
-        if past_k and past_v and past_k in graph_input_names and past_v in graph_input_names:
+
+        if past_k and past_v:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_attention.py:696]: if past_k and past_v:
    mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
elif key_padding_mask or add_qk:
    mha_inputs.extend([key_padding_mask, add_qk])

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        for output in skip_layernorm.output:
+            if not output:
+                continue
+            children = input_name_to_nodes[output]
+            children_types = [child.op_type for child in children]
+            if children_types.count("MatMul") >= 1:
+                root_input = output
+                break
+
+        # graph_input_names = set([node.name for node in self.model.graph().input])
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_conformer_attention.py:81]: remove commented code

File: onnxruntime/python/tools/transformers/optimizer.py [hunk 7]
```diff
@@ -32,6 +32,7 @@
 from onnx_model_bert_keras import BertOnnxModelKeras
 from onnx_model_bert_tf import BertOnnxModelTF
 from onnx_model_clip import ClipOnnxModel
+from onnx_model_ct import ConformerOnnxModel
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/optimizer.py:35]: rename onnx_model_ct to onnx_model_conformer

File: onnxruntime/python/tools/transformers/optimizer.py [hunk 8]
```diff
@@ -56,6 +57,7 @@
     "unet": (UnetOnnxModel, "pytorch", 1),  # UNet in Stable Diffusion
     "vae": (VaeOnnxModel, "pytorch", 1),  # UAE in Stable Diffusion
     "vit": (BertOnnxModel, "pytorch", 1),
+    "ct": (ConformerOnnxModel, "pytorch", 1),
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/optimizer.py:60]: rename ct to conformer

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+from fusion_attention import AttentionMask, FusionAttention
+from onnx_model import OnnxModel
+
+logger = logging.getLogger(__name__)
+
+
+class FusionConformerAttention(FusionAttention):
+    """
+    Fuse Conformer Attention subgraph into one Attention node.
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_conformer_attention.py:15]:
```suggestion
    Fuse Conformer Attention subgraph into one MultiHeadAttention node.
```

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        )
+
+        add_v = None
+        if v_nodes is not None:
+            (concat_v, _, _, add_v, matmul_v) = v_nodes
+            concat_parent = self.model.get_parent(concat_v, 0, None)
+            present_v = concat_v.output[0]
+            past_v = concat_parent.output[0]
+        else:
+            logger.debug("fuse_attention: failed to match v path")
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_conformer_attention.py:58]:
```suggestion
            logger.debug("fuse_conformer_attention: failed to match v path")
```
Can you also add similar `logger.debug` messages before the other `else` branches that have `return` that say what failed to match? When running the optimizer script with `--verbose`, these messages will print and can help with quickly finding any bugs.

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            past_k = concat_parent.output[0]
+            present_k = concat_k.output[0]
+        else:
+            return
+
+        attention_last_node = reshape_qkv
+        num_heads, hidden_size = self.get_num_heads_and_hidden_size(reshape_q)
+
+        if num_heads <= 0 or hidden_size <= 0 or (hidden_size % num_heads) != 0:
+            logger.debug("fuse_attention: failed to detect num_heads or hidden_size")
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_conformer_attention.py:97]:
```suggestion
            logger.debug("fuse_conformer_attention: failed to detect num_heads or hidden_size")
```

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            return
+
+        new_node = None
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_conformer_attention.py:100]: I think this can be removed since the next line redefines `new_node` and doesn't use `new_node` to redefine it. If `create_multihead_attention_node` fails, it should return `None`.

File: onnxruntime/test/python/transformers/conformer_model_generator.py [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+if __name__ == "__main__":
+    np.random.seed(2)
+    num_heads = 8
+    hidden_size = 512
+
+    model = create_conformer_attention(num_heads=num_heads, hidden_size=hidden_size)
+    onnx.save(model, "conformer_self_mha.onnx")
+
+    model = create_conformer_attention(num_heads=num_heads, hidden_size=hidden_size, fused=True)
+    onnx.save(model, "./test_data/models/conformer/conformer_self_mha_fused.onnx")
```
Comment (kunal-vaishnavi) [inline onnxruntime/test/python/transformers/conformer_model_generator.py:481]: You can update `cmake/onnxruntime_python.cmake` so that the CI pipelines can see this ONNX model.

There are several commands to add in the `cmake` file. Ex:
https://github.com/microsoft/onnxruntime/blob/02333293dec94922585a4aed39bd331128b643a6/cmake/onnxruntime_python.cmake#L436-L438
https://github.com/microsoft/onnxruntime/blob/02333293dec94922585a4aed39bd331128b643a6/cmake/onnxruntime_python.cmake#L550
https://github.com/microsoft/onnxruntime/blob/02333293dec94922585a4aed39bd331128b643a6/cmake/onnxruntime_python.cmake#L701-L703

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py (hunk 10 above)
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_conformer_attention.py:58]:
```suggestion
            logger.debug("fuse_conformer_attention: failed to match v path")
```
Can you also add similar `logger.debug` messages in the other `else` branches that have `return` that say what failed to match? When running the optimizer script with `--verbose`, these messages will print and can help with quickly finding any bugs.

File: onnxruntime/test/python/transformers/conformer_model_generator.py (hunk 13 above)
Comment (apsonawane) [reply onnxruntime/test/python/transformers/conformer_model_generator.py:543]: Done updated

File: onnxruntime/python/tools/transformers/fusion_conformer_attention.py (hunk 10 above)
Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_conformer_attention.py]: Updated

File: cmake/onnxruntime_python.cmake [hunk 14]
```diff
@@ -443,6 +443,9 @@ if (onnxruntime_BUILD_UNIT_TESTS)
   file(GLOB onnxruntime_python_transformers_testdata_whisper CONFIGURE_DEPENDS
       "${ONNXRUNTIME_ROOT}/test/python/transformers/test_data/models/whisper/*.onnx"
   )
+  file(GLOB onnxruntime_python_transformers_testdata_whisper CONFIGURE_DEPENDS
+      "${ONNXRUNTIME_ROOT}/test/python/transformers/test_data/models/conformer/*.onnx"
```
Comment (kunal-vaishnavi) [inline cmake/onnxruntime_python.cmake:447]: Rename to `onnxruntime_python_transformers_testdata_conformer`

File: cmake/onnxruntime_python.cmake [hunk 15]
```diff
@@ -711,6 +715,9 @@ if (onnxruntime_BUILD_UNIT_TESTS)
     COMMAND ${CMAKE_COMMAND} -E copy
         ${onnxruntime_python_transformers_testdata_whisper}
         $<TARGET_FILE_DIR:${build_output_target}>/transformers/test_data/models/whisper/
+    COMMAND ${CMAKE_COMMAND} -E copy
+        ${onnxruntime_python_transformers_testdata_whisper}
```
Comment (kunal-vaishnavi) [inline cmake/onnxruntime_python.cmake:719]: Rename to `onnxruntime_python_transformers_testdata_conformer`

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 16]
```diff
@@ -657,7 +657,6 @@ def create_multihead_attention_node(
             return None
 
         graph_input_names = set([node.name for node in self.model.graph().input])
-        graph_output_names = set([node.name for node in self.model.graph().output])
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_attention.py:660]: Why is this deleted?

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 17]
```diff
@@ -693,12 +692,15 @@ def create_multihead_attention_node(
...
+
+        if past_k and past_v:
             mha_inputs.extend([key_padding_mask, add_qk, past_k, past_v])
+        elif key_padding_mask or add_qk:
+            mha_inputs.extend([key_padding_mask, add_qk])
 
         # Add outputs for MHA
         mha_outputs = [output]
-        if present_k and present_v and present_k in graph_output_names and present_v in graph_output_names:
+        if present_k and present_v:
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/fusion_attention.py:703]: Since the Whisper test cases are failing in the CI pipelines, the checks that `present_k` and `present_v` are in `graph_output_names` may be needed

Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_attention.py:703]: They are failing in the CI pipeline because it is not able to find the file mentioned. graph_output_names is required because we are anyways appending present_k and present_v even though it is empty right?

CI pipeline is failing because of the above naming issue I think. Because locally these tests are passing, and earlier when conformer tests were failing these tests were passing so the issue with the cmake changes I made

File: cmake/onnxruntime_python.cmake (hunk 14 above)
Comment (apsonawane) [reply cmake/onnxruntime_python.cmake:447]: Updated

File: cmake/onnxruntime_python.cmake (hunk 15 above)
Comment (apsonawane) [reply cmake/onnxruntime_python.cmake:719]: Updated

File: onnxruntime/python/tools/transformers/fusion_attention.py (hunk 16 above)
Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_attention.py:660]: This is not required since we are not checking if present_k or present_v is there in the output names, we are always appending it directly to output

File: onnxruntime/python/tools/transformers/fusion_attention.py (hunk 17 above)
Comment (apsonawane) [reply onnxruntime/python/tools/transformers/fusion_attention.py:703]: Tests passed

**Review by kunal-vaishnavi — APPROVED** [review]
